test/MemorySizeTest.cpp

```diff
@@ -0,0 +1,213 @@
+// Copyright 2023, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (July of 2023,
+// schlegea@informatik.uni-freiburg.de)
+
+#include <absl/strings/str_cat.h>
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <ranges>
+#include <sstream>
+#include <vector>
+
+#include "util/ConstexprMap.h"
+#include "util/MemorySize/MemorySize.h"
+
+// Importing all the literals.
+using namespace ad_utility::memory_literals;
+
+// A small helper, because you can't call member functions on user defined
+// literals.
+size_t getBytes(const ad_utility::MemorySize& m) { return m.bytes(); };
+
+TEST(MemorySize, UserDefinedLiterals) {
+  // Normal bytes.
+  ASSERT_EQ(50uL, getBytes(50_B));
+
+  // Kilobytes.
+  ASSERT_EQ(2000uL, getBytes(2_kB));       // Whole number.
+  ASSERT_EQ(1500uL, getBytes(1.5_kB));     // Floating point without rounding.
+  ASSERT_EQ(1001uL, getBytes(1.0003_kB));  // Floating point with rounding.
+
+  // Megabytes.
+  ASSERT_EQ(2000000uL, getBytes(2_MB));    // Whole number.
+  ASSERT_EQ(1500000uL, getBytes(1.5_MB));  // Floating point without rounding.
+  ASSERT_EQ(1000001uL,
+            getBytes(1.0000003_MB));  // Floating point with rounding.
+
+  // Gigabytes.
+  ASSERT_EQ(2000000000uL, getBytes(2_GB));  // Whole number.
+  ASSERT_EQ(1500000000uL,
+            getBytes(1.5_GB));  // Floating point without rounding.
+  ASSERT_EQ(1000000001uL,
+            getBytes(1.0000000003_GB));  // Floating point with rounding.
+
+  // Terabytes.
+  ASSERT_EQ(2000000000000uL, getBytes(2_TB));  // Whole number.
+  ASSERT_EQ(1500000000000uL,
+            getBytes(1.5_TB));  // Floating point without rounding.
+  ASSERT_EQ(1000000000001uL,
+            getBytes(1.0000000000003_TB));  // Floating point with rounding.
+}
+
+// Describes a memory size in all available memory units.
+struct AllMemoryUnitSizes {
+  size_t bytes;
+  double kilobytes;
+  double megabytes;
+  double gigabytes;
+  double terabytes;
+};
+
+// A hash map pairing up a single memory size unit with the corresponding
+// `AllMemoryUnitSizes` representations.
+constexpr ad_utility::ConstexprMap<std::string_view, AllMemoryUnitSizes, 5>
+    singleMemoryUnitSizes(
+        {std::pair<std::string_view, AllMemoryUnitSizes>{
+             "B", AllMemoryUnitSizes{1uL, 1e-3, 1e-6, 1e-9, 1e-12}},
+         std::pair<std::string_view, AllMemoryUnitSizes>{
+             "kB", AllMemoryUnitSizes{1000uL, 1, 1e-3, 1e-6, 1e-9}},
+         std::pair<std::string_view, AllMemoryUnitSizes>{
+             "MB", AllMemoryUnitSizes{1000000uL, 1e3, 1, 1e-3, 1e-6}},
+         std::pair<std::string_view, AllMemoryUnitSizes>{
+             "GB", AllMemoryUnitSizes{1000000000uL, 1e6, 1e3, 1, 1e-3}},
+         std::pair<std::string_view, AllMemoryUnitSizes>{
+             "TB", AllMemoryUnitSizes{1000000000000uL, 1e9, 1e6, 1e3, 1}}});
```

Comment:
Okay, how many of the redundant typenames are actually needed? Get rid of the other ones, the types are clear. If some of them are necessary, add a short alias before them to make this more readable.
And use separators for the large numbers, so `1'000'000'000'000` for readability.

Reply:
That's the stupid thing: Only the types inside the `std::pair` were redundant. Doing something like `std::vector{...}`, or  `std::vector<...>{...}`, without any `std::pair` inside, just leads to compiler errors.

Tried to improve this in [76bb6c4](https://github.com/ad-freiburg/qlever/pull/1032/commits/76bb6c41214346554e16e7980105662a26766c7a) and [43a16d9](https://github.com/ad-freiburg/qlever/pull/1032/commits/43a16d9c33572992704c4b4114f8acd8af92e628).

Reply:
Yes, it's alright